src/lib/containers/provenance/ProvenanceGraphUtils.ts

```diff
@@ -0,0 +1,270 @@
+  entityHeaderIsCurrent: EntityHeaderIsCurrent
+  itemCount: number | undefined
+  nodesCopy: Node[]
+  edgesCopy: Edge[]
+}) => {
+  const { entityHeaderIsCurrent, itemCount = 0, nodesCopy, edgesCopy } = params
+  const expandNodeProps = {
+    type: NodeType.EXPAND,
+    data: {
+      itemCount,
+      entityHeaderIsCurrent,
+    },
+  }
+  const entityNodeProps = {
+    type: NodeType.ENTITY,
+    data: entityHeaderIsCurrent,
+  }
+  addNodeAndEdge({
+    newNodeProps: expandNodeProps,
+    existingNodeProps: entityNodeProps,
+    nodesCopy,
+    edgesCopy,
+  })
+}
+
+export const addEntityPlaceholderNode = (params: {
+  ref: Reference
+  activity: Activity
+  nodesCopy: Node[]
+  edgesCopy: Edge[]
+}) => {
+  const { ref, activity, nodesCopy, edgesCopy } = params
+  const entityPlaceholderNodeProps = {
+    type: NodeType.ENTITY_PLACEHOLDER,
+    data: ref,
+  }
+  const activityNodeProps = {
+    type: NodeType.ACTIVITY,
+    data: activity,
+  }
+  addNodeAndEdge({
+    newNodeProps: entityPlaceholderNodeProps,
+    existingNodeProps: activityNodeProps,
+    nodesCopy,
+    edgesCopy,
+  })
+}
+
+export const addUndefinedNode = (params: {
+  entityHeader: EntityHeader
+  nodesCopy: Node[]
+  edgesCopy: Edge[]
+}) => {
+  const { entityHeader, nodesCopy, edgesCopy } = params
+  const undefinedNodeProps = {
+    type: NodeType.UNDEFINED,
+    data: entityHeader,
+  }
+  const entityNodeProps = {
+    type: NodeType.ENTITY,
+    data: {
+      entityHeader,
+    },
+  }
+  addNodeAndEdge({
+    newNodeProps: undefinedNodeProps,
+    existingNodeProps: entityNodeProps,
+    nodesCopy,
+    edgesCopy,
+  })
+}
+
+export const addExternalNode = (params: {
+  usedURL: UsedURL
+  activity: Activity
+  nodesCopy: Node[]
+  edgesCopy: Edge[]
+}) => {
+  const { usedURL, activity, nodesCopy, edgesCopy } = params
+  const activityNodeProps = {
+    type: NodeType.ACTIVITY,
+    data: activity,
+  }
+  const externalNodeProps = {
+    type: NodeType.EXTERNAL,
+    data: usedURL,
+  }
+  addNodeAndEdge({
+    newNodeProps: externalNodeProps,
+    existingNodeProps: activityNodeProps,
+    nodesCopy,
+    edgesCopy,
+  })
+}
+
+export const addEntityNode = (params: {
+  entityHeaderIsCurrent: EntityHeaderIsCurrent
+  activity: Activity | undefined
+  nodesCopy: Node[]
+  edgesCopy: Edge[]
+}) => {
+  const { entityHeaderIsCurrent, activity, nodesCopy, edgesCopy } = params
+  const entityNodeProps = {
+    type: NodeType.ENTITY,
+    data: entityHeaderIsCurrent,
+  }
+  if (isNodeNotFound(entityNodeProps, nodesCopy)) {
+    // add the new entity node
+    nodesCopy.push(getProvenanceNode(entityNodeProps))
+    if (activity) {
+      const activityNodeProps = {
+        type: NodeType.ACTIVITY,
+        data: activity,
+      }
+      edgesCopy.push(getProvenanceEdge(entityNodeProps, activityNodeProps))
+    }
+  }
+}
+
+export const getEntityHeadersIsCurrent = async (params: {
+  refs: ReferenceList
+}): Promise<EntityHeaderIsCurrent[]> => {
```

Comment:
If the existing EntityHeader object suffices (as below), then you probably won't need the `EntityHeaderIsCurrent` custom data type.

Reply:
Absolutely